aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1335,81 +1335,487 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
+      // them into two arrays of intersection indices such that
+      // selected_src = repeat_interleave(src_int_idx, int_counts),
+      // and selected_sorted is obtained as follows:
+      // offsets = int_counts.cumsum(0).sub_(int_counts)
+      // for ii, (j, c) in enumerate(zip(sorted_int_idx, int_counts)):
+      //     out_slice = slice(offsets[ii], offsets[ii] + c)
+      //     src_slice = slice(j, j + c)
+      //     selected_sorted[out_slice] = sorted_int_idx[src_slice]
+      auto selected_sorted = at::empty({res_len}, sorted.options());
+      auto selected_src = at::empty({res_len}, src.options());
+
+      // fill in selected_sorted, selected_src
+      {
+        auto* ptr_selected_sorted = selected_sorted.data_ptr<int64_t>();
+        auto* ptr_selected_src = selected_src.data_ptr<int64_t>();
+
+        const auto thread_offsets = compressed_int_counts.cumsum(0).sub_(compressed_int_counts);
+        const auto* ptr_sorted_idx = sorted_idx.data_ptr<int64_t>();
+        at::parallel_for(0, src_len, src_grain_size, [&](int64_t start, int64_t end) {
+            const auto tid = at::get_thread_num();
+            const auto tid_offset = thread_offsets.data_ptr<int64_t>()[tid];
+            const auto* ptr_tid_src_int_idx = src_int_idx.select(0, tid).data_ptr<int64_t>();
+            const auto* ptr_tid_sorted_int_idx = sorted_int_idx.select(0, tid).data_ptr<int64_t>();
+            const auto* ptr_tid_int_counts = int_counts.select(0, tid).data_ptr<int64_t>();
+            auto* ptr_tid_selected_sorted = ptr_selected_sorted + tid_offset;
+            auto* ptr_tid_selected_src = ptr_selected_src + tid_offset;
+
+            for (const auto ii : c10::irange(start, end)) {
+              const auto count = *ptr_tid_int_counts++;
+              const auto i = *ptr_tid_src_int_idx++;
+              const auto j = *ptr_tid_sorted_int_idx++;
+              if (!count) continue;
+
+              std::fill_n(ptr_tid_selected_src, count, i);
+              std::copy_n(ptr_sorted_idx + j, count, ptr_tid_selected_sorted);
+
+              ptr_tid_selected_sorted += count;
+              ptr_tid_selected_src += count;
+            }
+        });
       }
-      for (const auto j : c10::irange(nnz)) {
-        int64_t jdx = cpu_dim_indices_ptr[j];
-        if (idx == jdx) {
-          new_nnz++;
-          iindices.push_back(i);
-          zindices.push_back(j);
+
+      return search_in_dim_indices
+        ? std::make_tuple(selected_sorted, selected_src)
+        : std::make_tuple(selected_src, selected_sorted);
+    };
+
+    // Converts a 1d sorted idx to a compressed 1d compressed idx,
+    // aka crow in the CSR format. Useful to get a count table in
+    // a parallelized and no-sync manner.
+    // TODO: this function could be very useful as a separate util.
+    const auto sorted_idx_to_cidx = [](
+        const Tensor& idx,
+        int64_t len,
+        bool run_in_parallel = true) -> Tensor {
```

Comment:
This the same function as in
https://github.com/pytorch/pytorch/blob/f26891c8b79a2fad1ab5f53cb5ddb355e870c03a/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp#L134-L162
I suggest using `_convert_indices_from_coo_to_csr` instead of `sorted_idx_to_cidx`.

Reply:
This case assumes nnz >> size, and the referenced implementation is not parallel.

Reply:
What do you mean that it is not parallel? The referenced implementation uses `at::paraller_for` similar to what is used here?
If `_convert_indices_from_coo_to_csr` is less performant than `sorted_idx_to_cidx`, but the functionality is the same, I'd recommend improving `_convert_indices_from_coo_to_csr` rather than introducing a new parallel implementation. In this way, the other tools using  `_convert_indices_from_coo_to_csr` would also benefit from this work.

Reply:
Sorry, missed that. I will test them. Will try to use this function directly hopefully without any header inclusion issues :)

Reply:
I will check whether parallel_for is inside a parallel-section-proof.

Reply:
Alas, this function is not public.